src/ui_nano_x.c

```diff
@@ -0,0 +1,235 @@
+      ""
+    });
+UX_STEP_CB(
+    ux_idle_flow_quit_step,
+    pb,
+    exit_app(),
+    {
+      &C_icon_dashboard_x,
+      "Quit",
+    });
+
+UX_FLOW(ux_idle_flow,
+    &ux_idle_flow_1_step,
+    &ux_idle_flow_cfg_step,
+    &ux_idle_flow_quit_step
+);
+
+// prompt
+#define PROMPT_SCREEN_NAME(idx) ux_prompt_flow_ ## idx ## _step
+#define PROMPT_SCREEN_TPL(idx) \
+    UX_STEP_NOCB( \
+        PROMPT_SCREEN_NAME(idx), \
+        bnnn_paging, \
+        { \
+            .title = G.prompt.screen[idx].prompt, \
+            .text = G.prompt.screen[idx].value, \
+        })
+
+PROMPT_SCREEN_TPL(0);
+PROMPT_SCREEN_TPL(1);
+PROMPT_SCREEN_TPL(2);
+PROMPT_SCREEN_TPL(3);
+PROMPT_SCREEN_TPL(4);
+PROMPT_SCREEN_TPL(5);
+PROMPT_SCREEN_TPL(6);
+
+static void prompt_response(bool const accepted) {
+    ui_initial_screen();
+    if (accepted) {
+        G.ok_callback();
+    } else {
+        G.cxl_callback();
+    }
+}
+
+UX_STEP_CB(
+    ux_prompt_flow_accept_step,
+    pb,
+    prompt_response(true),
+    {
+        &C_icon_validate_14,
+        "Accept"
+    });
+
+UX_STEP_CB(
+    ux_prompt_flow_reject_step,
+    pb,
+    prompt_response(false),
+    {
+        &C_icon_crossmark,
+        "Reject"
+    });
+
+UX_FLOW(ux_prompts_flow,
+    &PROMPT_SCREEN_NAME(0),
+    &PROMPT_SCREEN_NAME(1),
+    &PROMPT_SCREEN_NAME(2),
+    &PROMPT_SCREEN_NAME(3),
+    &PROMPT_SCREEN_NAME(4),
+    &PROMPT_SCREEN_NAME(5),
+    &PROMPT_SCREEN_NAME(6),
+    &ux_prompt_flow_reject_step,
+    &ux_prompt_flow_accept_step
+);
+_Static_assert(NUM_ELEMENTS(ux_prompts_flow) - 3 /*reject + accept + end*/ == MAX_SCREEN_COUNT, "ux_prompts_flow doesn't have the same number of screens as MAX_SCREEN_COUNT");
+
+void ui_cfg_screen(void) {
+
+    // reserve a display stack slot if none yet
+    if(G_ux.stack_count == 0) {
+        ux_stack_push();
+    }
+    ux_flow_init(0, ux_cfg_flow, NULL);
+
+}
+
+void ui_initial_screen(void) {
+
+    // reserve a display stack slot if none yet
+    if(G_ux.stack_count == 0) {
+        ux_stack_push();
+    }
+    ux_flow_init(0, ux_idle_flow, NULL);
+
+}
+
+__attribute__((noreturn))
+void ui_prompt(const char *const *labels, ui_callback_t ok_c, ui_callback_t cxl_c) {
+    check_null(labels);
+
+    size_t const screen_count = ({
+        size_t i = 0;
+        while (i < MAX_SCREEN_COUNT && labels[i] != NULL) { i++; }
+        i;
+    });
+
+    // We fill the destination buffers at the end instead of the beginning so we can
+    // use the same array for any number of screens.
+    size_t const offset = MAX_SCREEN_COUNT - screen_count;
+    for (size_t i = 0; labels[i] != NULL; i++) {
+        const char *const label = (const char *)PIC(labels[i]);
+        if (strlen(label) > sizeof(G.prompt.screen[i].prompt)) THROW(EXC_MEMORY_ERROR);
+        strcpy(G.prompt.screen[offset + i].prompt, label);
+
+        G.prompt.callbacks[i](G.prompt.screen[offset + i].value, sizeof(G.prompt.screen[offset + i].value), G.prompt.callback_data[i]);
```

Comment:
Does sizeof work correctly here? I may have this wrong but I thought if .value is a pointer, it would just be `sizeof(G.prompt.screen[offset + i].value) == 1`.

Reply:
G.prompt.screen[offset+i].value is an actual array in a struct, not a pointer, and we're accessing it here without letting it decay into a pointer, so sizeof returns the size of the array. (It'd also be sizeof(ptr) = 4, I think, but that's not the point here.)